Add TLS settings to all connection settings (#205)
open-telemetry/opamp-spec: specification.md

connection credentials. This way individual Agent's credentials may be revoked
without disrupting the access to all other Agents.

#### Agent-initiated CA trust Flow

Status: [Development]

> This is an Agent-initiated flow that allows an agent to connect to a Server without

Author: Right now this does not use an explicit signal (from the agent to the server).

We can add one as a part of the `OpAMPConnectionSettingsRequest` message.
But then I'm concerned that we don't have a way for the agent to signal to the server that offered settings have been attempted, but have failed.

connection credentials. This way individual Agent's credentials may be revoked
without disrupting the access to all other Agents.

#### Agent-initiated CA trust Flow

Status: [Development]

This is an Agent-initiated flow that allows an agent to connect to a Server

> without trusting its CA. Initially the agent will connect without any form of

Reviewer: It is not really "without trusting". The Agent will trust the CA once the flow is completed. It is just that the Agent does not have the CA certificate initially. Perhaps reword it like this?
```suggestion
without have the possession of the Server's CA initially. Initially the agent will connect without any form of
```

verification, the initial AgentToServer message will indicate that the agent

> expects connection connection settings to be offered in the response message.

Reviewer: ```suggestion
expects connection settings to be offered in the response message.
```

> This is an Agent-initiated flow that allows an agent to connect to a Server
> without trusting its CA. Initially the agent will connect without any form of
> verification, the initial AgentToServer message will indicate that the agent
> expects connection connection settings to be offered in the response message.
> The response contains the PEM content associated with its CA certificate as part
> of the ServerToAgent response using the connection settings offering.

Reviewer 2: ```suggestion
This is an Agent-initiated flow that allows an agent to connect to a Server
without trusting its CA. Initially the agent will connect without any form of
verification. The initial AgentToServer message will indicate using SettingsRequest that the agent
expects connection settings to be offered in the response message.
The response contains the PEM content associated with its CA certificate as part
of the ServerToAgent response using the connection settings offering.
```

connection credentials. This way individual Agent's credentials may be revoked
without disrupting the access to all other Agents.

#### Agent-initiated CA trust Flow

Status: [Development]

This is an Agent-initiated flow that allows an agent to connect to a Server

> without the possesion of the Server's CA initially. Initially the agent will

Reviewer: BTW, I think this same flow can also work for _replacing_ the CA. So even if the Agent knows a CA that the Server uses initially, the Server can offer a different CA to be used in the future. The difference will be that the initial connection will be with CA verification. Do we want to generalize this flow to also allow this use case?

```protobuf
message ConnectionSettingsRequest {
    OpAMPConnectionSettingsRequest opamp = 1;
    SettingsRequest settings_request = 2;
}
```

The `opamp` field is set to indicate a request for OpAMP connection settings.

> The `settings_request` field is set to indicate a request for all offered

Reviewer: "for all" or specifically "for TLSConnectionSettings"? What does "all" mean in this context?

Author: All settings for OpAMP  connection (this may include the hosts, and things like the heartbeat)

Other connection settings. These are Agent-specific and are up to the Agent
interpret.

##### OtherConnectionSettings.tls

Status: [Development]

Optional connection specific TLS settings.

#### TLSConnectionSettings Message

Status: [Development]

The message carries optional TLS settings that are used to configure a client's
connection. If the Agent is able to validate the connection settings, the Agent
SHOULD forget any previous TLS settings. If this message is not included, the

> client SHOULD use the agent's default TLS settings for the connection.

Author: @tigrannajaryan @andykellr, I've added a description on how TLS settings should function, and have changed the demo (https://github.com/open-telemetry/opamp-go/pull/338) to reflect this.

The goal is that when `TLSConnectionSettings` is present, it can form the TLS config (excluding client certificate).
If no TLS settings are sent with a connection offering, the agent's defaults should be used.
